File: app/src/main/java/com/anadi/weatherinfo/addlocation/AddLocationPresenter.java
package com.anadi.weatherinfo.addlocation;

import android.os.Handler;
import android.text.TextUtils;

import com.anadi.weatherinfo.R;
import com.anadi.weatherinfo.repository.Locations;
import com.anadi.weatherinfo.repository.LocationsCash;

import java.util.ArrayList;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import timber.log.Timber;

public class AddLocationPresenter implements AddLocationContract.Presenter {

    private final AddLocationContract.View view;
    private final AddLocationContract.Model model;
    private final LocationsProvider locations;
    private final Handler handler = new Handler();

    private final Executor exec;

    AddLocationPresenter(AddLocationContract.View view) {
        this.view = view;
        model = LocationsCash.getInstance();
        locations = Locations.getInstance();
        exec = Executors.newSingleThreadExecutor();
    }

    public void addLocation(final String selectedCity, final String selectedCountry) {
        view.loading();

        if (TextUtils.isEmpty(selectedCity) || TextUtils.isEmpty(selectedCountry) || selectedCity.equalsIgnoreCase(
                "Select Item") || selectedCountry
                .equalsIgnoreCase("Select Item")) {
            Timber.d("selectedCity: " + selectedCity + "selectedCountry: " + selectedCountry);
            view.onError(R.string.on_error_select_city);
            return;
        }

        exec.execute(() -> {
            try {
                boolean result = model.add(selectedCity, selectedCountry);

                if (result) {
                    onCityAdded();
                } else {
                    onError();
                }

            } catch (Exception e) {
                System.err.println(e.getMessage());
                e.printStackTrace();
            }
        });
    }

    @Override
    public void onCountrySelected(String countryName) {
        view.updateCityList(locations.getCityNames(countryName));
    }

    @Override
    public ArrayList<String> getCountryNames() {
        return locations.getCountryNames();
    }

    private void onCityAdded() {
        handler.post(() -> view.onAddedSuccess());
    }

    private void onError() {
        handler.post(() -> view.onError(R.string.on_error_select_city));
    }
}

File: app/src/main/java/com/anadi/weatherinfo/addlocation/AddLocationPresenter.kt
package com.anadi.weatherinfo.addlocation

import android.os.Handler
import android.text.TextUtils
import com.anadi.weatherinfo.R
import com.anadi.weatherinfo.repository.Locations
import com.anadi.weatherinfo.repository.LocationsCash
import timber.log.Timber
import java.util.ArrayList
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class AddLocationPresenter internal constructor(private val view: AddLocationContract.View) : AddLocationContract.Presenter {
    private val model: AddLocationContract.Model
    private val locations: LocationsProvider
    private val handler = Handler()
    private val exec: Executor
    override fun addLocation(selectedCity: String?, selectedCountry: String?) {
        view.loading()
        if (TextUtils.isEmpty(selectedCity) || TextUtils.isEmpty(selectedCountry) || selectedCity.equals(
                        "Select Item", ignoreCase = true) || selectedCountry
                        .equals("Select Item", ignoreCase = true)) {
            Timber.d("selectedCity: " + selectedCity + "selectedCountry: " + selectedCountry)
            view.onError(R.string.on_error_select_city)
            return
        }
        exec.execute {
            try {
                val result = model.add(selectedCity, selectedCountry)
                if (result) {
                    onCityAdded()
                } else {
                    onError()
                }
            } catch (e: Exception) {
                System.err.println(e.message)
                e.printStackTrace()
            }
        }
    }

    override fun onCountrySelected(countryName: String?) {
        view.updateCityList(locations.getCityNames(countryName)!!)
    }

    override val countryNames: ArrayList<String?>?
        get() = locations.countryNames

    private fun onCityAdded() {
        handler.post { view.onAddedSuccess() }
    }

    private fun onError() {
        handler.post { view.onError(R.string.on_error_select_city) }
    }

    init {
        model = LocationsCash.getInstance()
        locations = Locations.getInstance()
        exec = Executors.newSingleThreadExecutor()
    }
}